### PR TITLE
Packed External Nullifier repr

### PR DESCRIPTION
**Pull request by Dzejkop**

### CODE REVIEW

File: world-chain-builder/crates/world/pbh/src/external_nullifier.rs [hunk 1]
```diff
@@ -1,121 +1,155 @@
...
-    InvaldFormat { actual: String },
-
-    #[error("error parsing prefix - {0}")]
-    InvalidPrefix(strum::ParseError),
-
-    #[error("error parsing date marker - {0}")]
-    InvalidDateMarker(#[from] DateMarkerParsingError),
-
-    #[error("error parsing nonce - {0}")]
-    InvalidNonce(std::num::ParseIntError),
```
Comment (0xForerunner) [inline world-chain-builder/crates/world/pbh/src/external_nullifier.rs]: LGTM! Only suggestion would be to leave in the granular errors! Might be nice to have.

**Review by 0xForerunner — APPROVED** [review]
Looks good!

**Review by 0xOsiris — APPROVED** [review]
LGTM, nice work!

File: contracts/src/helpers/PBHExternalNullifier.sol [hunk 2]
```diff
@@ -25,25 +30,33 @@ library PBHExternalNullifier {
...
 
+    uint8 public constant V1 = 1;
+
     /// @notice Encodes a PBH external nullifier using the provided year, month, and nonce.
     /// @param pbhNonce An 8-bit nonce value (0-255) used to uniquely identify the nullifier within a month.
     /// @param month An 8-bit 1-indexed value representing the month (1-12).
     /// @param year A 16-bit value representing the year (e.g., 2024).
     /// @return The encoded PBHExternalNullifier.
-    function encode(uint8 pbhNonce, uint8 month, uint16 year) internal pure returns (uint256) {
+    function encode(uint8 version, uint8 pbhNonce, uint8 month, uint16 year) internal pure returns (uint256) {
```
Comment (0xKitsune) [inline contracts/src/helpers/PBHExternalNullifier.sol:44]: If `V1` is constant, we can probably omit `version` from the args and encode the external nullifier with `V1` directly.

File: contracts/src/helpers/PBHExternalNullifier.sol [hunk 3]
```diff
@@ -52,7 +65,8 @@ library PBHExternalNullifier {
     /// @dev This function ensures the external nullifier matches the current year and month,
     ///      and that the nonce does not exceed `numPbhPerMonth`.
     function verify(uint256 externalNullifier, uint8 numPbhPerMonth) public view {
-        (uint8 pbhNonce, uint8 month, uint16 year) = PBHExternalNullifier.decode(externalNullifier);
+        (uint8 version, uint8 pbhNonce, uint8 month, uint16 year) = PBHExternalNullifier.decode(externalNullifier);
```
Comment (0xForerunner) [inline contracts/src/helpers/PBHExternalNullifier.sol:68]: Found an issue here! We're actually under constrained. We need to make sure that the leading zeros here are as expected.
```suggestion
        require(externalNullifier <= type(uint40).max, InvalidExternalNullifierLeadingZeros());
```

File: contracts/src/helpers/PBHExternalNullifier.sol [hunk 4]
```diff
@@ -52,7 +65,8 @@ library PBHExternalNullifier {
     /// @dev This function ensures the external nullifier matches the current year and month,
     ///      and that the nonce does not exceed `numPbhPerMonth`.
     function verify(uint256 externalNullifier, uint8 numPbhPerMonth) public view {
```
Comment (0xForerunner) [inline contracts/src/helpers/PBHExternalNullifier.sol:73]: Found an issue here! We're actually under constrained. We need to make sure that the leading zeros here are as expected.
```suggestion
    function verify(uint256 externalNullifier, uint8 numPbhPerMonth) public view {
     require(externalNullifier <= type(uint40).max, InvalidExternalNullifierLeadingZeros());
```

File: contracts/src/helpers/PBHExternalNullifier.sol [hunk 5]
```diff
@@ -8,11 +8,16 @@ import "@BokkyPooBahsDateTimeLibrary/BokkyPooBahsDateTimeLibrary.sol";
...
-///      - Bits 32-255: Empty
-///      - Bits 16-31: Year
-///      - Bits 8-15: Month
-///      - Bits 0-7: Nonce
+///      - Bits:40-255: Empty
+///      - Bits 32-39: Year
+///      - Bits 16-31: Month
+///      - Bits 8-15: Nonce
+///      - Bits 0-7: Version
 library PBHExternalNullifier {
```
Comment (0xForerunner) [inline contracts/src/helpers/PBHExternalNullifier.sol:16]:
```suggestion
library PBHExternalNullifier {
    /// @notice Thrown when the provided external nullifier doesn't
    /// contain the correct leading zeros
    error InvalidExternalNullifierLeadingZeros();
    
```

File: contracts/src/helpers/PBHExternalNullifier.sol (hunk 4 above)
Comment (Dzejkop) [reply contracts/src/helpers/PBHExternalNullifier.sol:73]: Thanks!

File: world-chain-builder/crates/world/pbh/src/external_nullifier.rs (hunk 1 above)
Comment (Dzejkop) [reply world-chain-builder/crates/world/pbh/src/external_nullifier.rs]: Good point I'll re-add month validation

Comment (Dzejkop) [reply world-chain-builder/crates/world/pbh/src/external_nullifier.rs]: Other errors don't really make sense at "parse time"

File: contracts/src/helpers/PBHExternalNullifier.sol (hunk 2 above)
Comment (Dzejkop) [reply contracts/src/helpers/PBHExternalNullifier.sol:44]: it's true but it makes the API asymmetric and makes testing a little uglier. I could add an `encodeV1` function though?

Comment (Dzejkop) [reply contracts/src/helpers/PBHExternalNullifier.sol:44]: I'll leave it as is. I'm open to changing it later

**Pull request merged**